examples/nlp/md/text_generation_with_miniature_gpt.md

```diff
@@ -0,0 +1,668 @@
+        self.dropout1 = layers.Dropout(rate)
+        self.dropout2 = layers.Dropout(rate)
+
+    def call(self, inputs):
+        attention_output = self.att(inputs)
+        attention_output = self.dropout1(attention_output)
+        out1 = self.layernorm1(inputs + attention_output)
+        ffn_output = self.ffn(out1)
+        ffn_output = self.dropout2(ffn_output)
+        return self.layernorm2(out1 + ffn_output)
+
+
+```
+
+---
+## Implement embedding layer
+
+Two seperate embedding layers, one for tokens, one for token index (positions).
+
+
+
+```python
+
+class TokenAndPositionEmbedding(layers.Layer):
+    def __init__(self, maxlen, vocab_size, emded_dim):
+        super(TokenAndPositionEmbedding, self).__init__()
+        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=emded_dim)
+        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=emded_dim)
+
+    def call(self, x):
+        maxlen = tf.shape(x)[-1]
+        positions = tf.range(start=0, limit=maxlen, delta=1)
+        positions = self.pos_emb(positions)
+        x = self.token_emb(x)
+        return x + positions
+
+
+```
+
+---
+## Implement miniature GPT model
+
+
+
+```python
+vocab_size = 20000  # Only consider the top 20k words
+maxlen = 20  # Max sequence size
+embed_dim = 64  # Embedding size for each token
+num_heads = 2  # Number of attention heads
+feed_forward_dim = 64  # Hidden layer size in feed forward network inside transformer
+
+
+def create_model():
+    inputs = layers.Input(shape=(maxlen,), dtype=tf.int32)
+    embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
+    x = embedding_layer(inputs)
+    transformer_block = TransformerBlock(embed_dim, num_heads, feed_forward_dim)
+    x = transformer_block(x)
+    outputs = layers.Dense(vocab_size)(x)
+    model = keras.Model(inputs=inputs, outputs=[outputs, x])
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    model.compile(
+        "adam", loss=[loss_fn, None],
+    )  # No loss and optimization based on word embeddings from transformer block
+    return model
+
+
+```
+
+---
+## Prepare data for word level language modelling
+
+We will download IMDB data, and combine training and validation sets for
+text generation task.
+
+
+
+```python
+!curl -O https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
+!tar -xf aclImdb_v1.tar.gz
+
+```
+
+```python
+
+batch_size = 32
+
+# The dataset contains each review in a separate text file
+# The text files are present in four different folders
+# Create a list all files
+filenames = []
+directories = [
+    "aclImdb/train/pos",
+    "aclImdb/train/neg",
+    "aclImdb/test/pos",
+    "aclImdb/test/neg",
+]
+for dir in directories:
+    for f in os.listdir(dir):
+        filenames.append(os.path.join(dir, f))
+
+print(f"{len(filenames)} files")
+
+# Create dataset from text files
+text_ds = tf.data.TextLineDataset(filenames)
```

Comment:
One thing that isn't clear to me with the use of `TextLineDataset` is this:
is one sample a *line* from a file, or the entire content of a file regardless of line breaks? I presume files may contain line breaks. It would be useful to print a few reviews from the dataset to check what they look like. Doing so would also have pointed you towards stripping the `<br>` tags in the first place.

Reply:
`TextLineDataset` creates concatenation of all the lines from all the files. If file 1 has 2 lines, and file 2 has 3 lines, the dataset from `[file1, file2]` will have 5 items. 
Incase of IMDB, all the files seem to have just one line. The line breaks are present in the form of `<br>` tag.
I noticed them but I thought its fine for demonstration purposes even if we keep them. Or if we print them as `\n`. For example, we are also ignoring periods and commas which makes input text somewhat meaningless by joining all the sentences.